### DISCUSSION TITLE
Azure function(javascript) - read local.settings.json property during testing

### DISCUSSION THREAD

**Discussion by jason-mwnz** [Azure/azure-functions-nodejs-worker#475] [Q&A] [answered]
- nodejs 14

- JavaScript

During the development time, i can use "process.env[propertyName]" to access the property value inside the local.settings.json file fine. Then i install Jest framework to do the unit testing, i just realised my function app failed to read any values from local.setting.json, or it actually running different environment during development and unit test(i print out the all the value from process.env object during unit testing and development and compared, they are quite different.)

Q1 
How to correct load the local.setting.json during unit testing phrase ?

Cheers

**Comment by v-bbalaiagar** [comment]
Transferring this issue to core tools for further insights.

**Comment by ejizba** [comment, marked as answer]
Hi @jason-mwnz the "local.settings.json" file is only used when running locally with the func core tools cli. For any other scenario, including unit tests, it will not be used. One reason is that the file should not be tracked in git because it may contain secrets, meaning it will not be useable in CI. If you can describe how you run your tests locally (from VS Code, the CLI) and/or from CI (Azure Pipelines, GitHub Actions, etc.) I might be able to help give a solution

We have https://github.com/Azure/azure-functions-nodejs-worker/issues/120 to track better docs on testing practices and otherwise I thought I'd try converting this to a "discussion" (never tried this before) since there won't be any action item on our side

**Reply by jason-mwnz** [reply]
@ejizba so if like what you say the "local.settings.json" is only works when running locally, but then what is the proper way or solution to do the unit testing for the azure function ? 

I install the "jest" package and write quite few unit tests for our function app. Then we have a Unit testing task during the CICD pipeline to trigger to run the test before it deploy to azure.

**Reply by ejizba** [reply]
What are you using to run CICD? And what are the environment variables for - are they connection strings, flags, or something else?

**Reply by jason-mwnz** [reply]
@zhajie we are using azure DevOps.  Some of variables for flags, some of them are required to generate a file, And some of value are require for expire time etc....

**Reply by ejizba** [reply]
Azure DevOps has support for user defined variables, which is what I would suggest: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#user-defined-variables

If there aren't any secrets, you can put the variables directly in your yaml. If it does have secrets you'll want to use the UI so that it can be encrypted and set as secret.

Lastly, you might want to check out jest's `beforeEach` if you want different values for different test suites.
